File: app/src/main/java/ua/hospes/rtm/ui/MainActivity.java
package ua.hospes.rtm.ui;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.ftinc.scoop.Scoop;

import dagger.android.support.DaggerAppCompatActivity;
import dagger.android.support.HasSupportFragmentInjector;
import ua.hospes.rtm.R;
import ua.hospes.rtm.core.StopWatchService;
import ua.hospes.rtm.ui.cars.CarsFragment;
import ua.hospes.rtm.ui.drivers.DriversFragment;
import ua.hospes.rtm.ui.race.RaceFragment;
import ua.hospes.rtm.ui.settings.SettingsFragment;
import ua.hospes.rtm.ui.teams.TeamsFragment;

public class MainActivity extends DaggerAppCompatActivity implements HasSupportFragmentInjector, NavigationView.OnNavigationItemSelectedListener {
    private DrawerLayout mDrawerLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Scoop.getInstance().apply(this);
        setContentView(R.layout.activity_main);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mDrawerLayout = findViewById(R.id.drawer_layout);
        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, mDrawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        mDrawerLayout.addDrawerListener(toggle);
        toggle.syncState();


        if (savedInstanceState == null) {
            navigationView.setCheckedItem(R.id.nav_race);
            getSupportFragmentManager().beginTransaction()
                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                    .replace(R.id.container, RaceFragment.newInstance())
                    .commit();
        }

        StopWatchService.checkDeath(this);
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == 999) recreate();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                mDrawerLayout.openDrawer(GravityCompat.START);
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        // set item as selected to persist highlight
        menuItem.setChecked(true);
        // close drawer when item is tapped
        mDrawerLayout.closeDrawers();

        // Add code here to update the UI based on the item selected
        // For example, swap UI fragments here

        switch (menuItem.getItemId()) {
            case R.id.nav_race:
                getSupportFragmentManager().beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .replace(R.id.container, RaceFragment.newInstance())
                        .commit();
                break;

            case R.id.nav_teams:
                getSupportFragmentManager().beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .replace(R.id.container, TeamsFragment.newInstance())
                        .commit();
                break;

            case R.id.nav_drivers:
                getSupportFragmentManager().beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .replace(R.id.container, DriversFragment.newInstance())
                        .commit();
                break;

            case R.id.nav_cars:
                getSupportFragmentManager().beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .replace(R.id.container, CarsFragment.newInstance())
                        .commit();
                break;

            case R.id.nav_settings:
                getSupportFragmentManager().beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .replace(R.id.container, SettingsFragment.newInstance())
                        .commit();
                break;

            default:
                getSupportFragmentManager().beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .replace(R.id.container, PlaceholderFragment.newInstance())
                        .commit();
                break;
        }

        return true;
    }


    @Override
    public void onBackPressed() {
        if (mDrawerLayout.isDrawerOpen(GravityCompat.START))
            mDrawerLayout.closeDrawer(GravityCompat.START);
        else
            super.onBackPressed();
    }


    public static class PlaceholderFragment extends Fragment {
        public PlaceholderFragment() {}

        public static PlaceholderFragment newInstance() {
            return new PlaceholderFragment();
        }

        @Override
        public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.fragment_main, container, false);
        }
    }
}

File: app/src/main/java/ua/hospes/rtm/ui/MainActivity.kt
package ua.hospes.rtm.ui

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import androidx.fragment.app.Fragment
import com.google.android.material.navigation.NavigationView
import dagger.hilt.android.AndroidEntryPoint
import ua.hospes.rtm.R
import ua.hospes.rtm.core.StopWatchService
import ua.hospes.rtm.databinding.ActivityMainBinding
import ua.hospes.rtm.domain.preferences.PreferencesManager
import ua.hospes.rtm.ui.cars.CarsFragment
import ua.hospes.rtm.ui.drivers.DriversFragment
import ua.hospes.rtm.ui.race.RaceFragment
import ua.hospes.rtm.ui.settings.SettingsFragment
import ua.hospes.rtm.ui.teams.TeamsFragment
import ua.hospes.rtm.ui.test.ComposeTestFragment
import ua.hospes.rtm.utils.extentions.doOnApplyWindowInsets
import javax.inject.Inject

private const val REQUEST_CODE_PRIVACY = 10
private const val REQUEST_CODE_TERMS = 20

@AndroidEntryPoint
class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    @Inject lateinit var prefs: PreferencesManager
    private val binding by lazy { ActivityMainBinding.inflate(layoutInflater) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        WindowCompat.setDecorFitsSystemWindows(window, false)
        binding.toolbar.doOnApplyWindowInsets { view, insets, padding, _ ->
            view.updatePadding(top = padding.top + insets.getInsets(WindowInsetsCompat.Type.systemBars()).top)
        }
        setSupportActionBar(binding.toolbar)

        binding.navView.setNavigationItemSelectedListener(this)

        val toggle = ActionBarDrawerToggle(this, binding.drawerLayout, binding.toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        binding.drawerLayout.addDrawerListener(toggle)
        toggle.syncState()


        if (savedInstanceState == null) {
            binding.navView.setCheckedItem(R.id.nav_race)
            navigateTo(RaceFragment())
        }

        if (!prefs.isPrivacyAccepted) startActivityForResult(Intent(this, PrivacyActivity::class.java), REQUEST_CODE_PRIVACY)
        else if (!prefs.isTermsAccepted) startActivityForResult(Intent(this, TermsActivity::class.java), REQUEST_CODE_TERMS)

        StopWatchService.checkDeath(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        android.R.id.home -> binding.drawerLayout.openDrawer(GravityCompat.START).let { true }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
        binding.drawerLayout.closeDrawers()

        when (menuItem.itemId) {
            R.id.nav_race -> navigateTo(RaceFragment()).also { menuItem.isChecked = true }
            R.id.nav_teams -> navigateTo(TeamsFragment()).also { menuItem.isChecked = true }
            R.id.nav_drivers -> navigateTo(DriversFragment()).also { menuItem.isChecked = true }
            R.id.nav_cars -> navigateTo(CarsFragment()).also { menuItem.isChecked = true }
            R.id.nav_settings -> navigateTo(SettingsFragment()).also { menuItem.isChecked = true }
            R.id.nav_test -> navigateTo(ComposeTestFragment()).also { menuItem.isChecked = true }

            R.id.nav_privacy -> startActivity(Intent(this, PrivacyActivity::class.java))
            R.id.nav_terms -> startActivity(Intent(this, TermsActivity::class.java))

            else -> Unit
        }
        return true
    }

    override fun onBackPressed() {
        if (binding.drawerLayout.isDrawerOpen(GravityCompat.START))
            binding.drawerLayout.closeDrawer(GravityCompat.START)
        else
            super.onBackPressed()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) = when (requestCode) {
        REQUEST_CODE_PRIVACY -> when (resultCode) {
            Activity.RESULT_OK ->
                if (!prefs.isTermsAccepted)
                    startActivityForResult(Intent(this, TermsActivity::class.java), REQUEST_CODE_TERMS)
                else Unit
            else -> finish()
        }
        REQUEST_CODE_TERMS -> when (resultCode) {
            Activity.RESULT_OK -> Unit
            else -> finish()
        }
        else -> super.onActivityResult(requestCode, resultCode, data)
    }

    private fun navigateTo(fragment: Fragment) = supportFragmentManager.beginTransaction()
            .replace(R.id.container, fragment)
            .commit()
}